src/main/java/com/snowflake/kafka/connector/internal/streaming/TopicPartitionChannel.java

```diff
@@ -131,19 +146,31 @@ public void insertRecordToBuffer(SinkRecord record) {
     }
   }
 
-  /* TODO: SNOW-529753 Deal with rebalances and restarts */
-  /* TODO: SNOW-536429 For EOS, also fetch the offset from snowflake during rebalance/restart */
-  private void init(long recordOffset) {
+  /* For EOS, fetch the offset from snowflake during only during first time after connector starts and record is received in Kafka Connector */
+  @VisibleForTesting
+  protected void fetchLastCommittedOffsetToken() {
+    String offsetToken = this.channel.getLatestCommittedOffsetToken();
     try {
-      startCleaner(recordOffset);
-    } catch (Exception e) {
-      LOGGER.error("Cleaner and Flusher threads shut down before initialization");
+      if (offsetToken == null) {
+        this.offsetPersistedInSnowflake = NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE;
+      } else {
+        this.offsetPersistedInSnowflake = Long.parseLong(offsetToken);
+      }
```

Comment:
Nit, this can be consolidated to `this.offsetPersistedInSnowflake = offsetToken == null ? NO_OFFSET_TOKEN_REGISTERED_IN_SNOWFLAKE : Long.parseLong(offsetToken);`

Reply:
Will do, however, I will handle that as part of https://github.com/snowflakedb/snowflake-kafka-connector/pull/395/commits (Which is still in draft) That PR refactors some of the common code paths.